src/storage/src/storage_state.rs

```diff
@@ -92,6 +96,79 @@ pub struct StorageState {
     /// Frontier of sink writes (all subsequent writes will be at times at or
     /// equal to this frontier)
     pub sink_write_frontiers: HashMap<GlobalId, Rc<RefCell<Antichain<Timestamp>>>>,
+    /// See: [SinkHandle]
+    pub sink_handles: HashMap<GlobalId, SinkHandle>,
+}
+
+/// This maintains an additional read hold on the source data for a sink, alongside
+/// the controller's hold and the handle used to read the shard internally.
+/// This is useful because environmentd's hold might expire, and the handle we use
+/// to read advances ahead of what we've successfully committed.
+/// In theory this could be stored alongside the other sink data, but this isn't
+/// intended to be a long term solution; either this should become a "critical" handle
+/// or environmentd will learn to hold its handles across restarts and this won't be
+/// needed.
+pub struct SinkHandle {
+    downgrade_tx: UnboundedSender<Antichain<Timestamp>>,
+    _handle: JoinHandle<()>,
+}
+
+impl SinkHandle {
+    /// A new handle.
+    pub fn new(
+        persist_location: PersistLocation,
+        shard_id: ShardId,
+        persist_clients: Arc<Mutex<PersistClientCache>>,
+    ) -> SinkHandle {
+        let (downgrade_tx, mut rx) = mpsc::unbounded_channel();
+
+        let _handle = mz_ore::task::spawn(|| "Sink handle advancement", async move {
+            let client = persist_clients
+                .lock()
+                .await
+                .open(persist_location)
+                .await
+                .expect("opening persist client");
+
+            let mut read_handle: ReadHandle<SourceData, (), Timestamp, Diff> = client
+                .open_reader(shard_id)
+                .await
+                .expect("opening reader for shard");
+
+            let mut downgrade_to = read_handle.since().clone();
```

Comment:
we believe this to be the shard-global since right? Is this not equivalent to the `StorageSinkDesc`'s `as_of`? perhaps ignore this comment, as this fix is temporary

Reply:
Yes; probably most of the time; and I think it doesn't matter too much as long as we're not incorrectly advancing the handle, so this seems like the easy thing at least!